stack trace tutorial

syzygyzer opened the pull request:
This is an in-depth examination of what the kernel is doing when you run a stack
trace on `-time ~s1`. This PR also includes a link to the tutorial from arvo.md,
which is practically a prerequisite for following this tutorial.

Lingering issues:

-Placement-
 
I'm not sure this is the correct place for this. It was originally part of
arvo.md, since knowing how the kernel is structured is essential, but it also
seems to be too hidden there. Every developer will run a stack trace at some
point, so this needs to be highly visible.

I've put it in hoon-errors.md for now, but I'm not totally happy with that
either. It's located in the reference section, which IMO is not the right place
for a tutorial.

Putting it somewhere in the Hoon tutorial makes some sense, and that is where
the bulk of our tutorials are located.

I think there is an argument for giving it its own page as well, but I'm not
sure where.
 
-Tables-

There's no line breaks for markdown tables, making the ducts harder to read. Do
we have support for HTML tables? @matildepark

The tables are also really wide. If anyone has suggestions on how to make them
more readable I'm all ears.

-Jargon-

I tried to be very careful about using the right words - card, move, gift, task,
etc - but its quite possible I messed something up as I ended up saying card
when I meant move etc several times in the rough drafts. I'd appreciate it if whoever
reviews this pays close attention to whether I am using those words correctly. 

-Soon to be added-

I'm whipping up a rather involved diagram showing the entire stack trace
graphically, including steps outside of the main pass->give sequence. I should
have it done early next week. I'll include it as part of this PR if its still
open by then but the doc is still solid without it so that's why I'm submitting now.

----

#

matildepark commented:
>There's no line breaks for markdown tables, making the ducts harder to read. Do
we have support for HTML tables? @matildepark

> The tables are also really wide. If anyone has suggestions on how to make them
more readable I'm all ears.

Were you testing this in a local urbit.org environment? Just curious.

Yes, you can use HTML tables if you like, it will work fine. You just can't mingle Markdown and HTML markup styles in the same block of text, really.

syzygyzer commented:
> Were you testing this in a local urbit.org environment? Just curious.

No, I haven't tried to set that up yet. Mostly because how to do so consists entirely of unknown unknowns for me, so I have no idea how long it will take and I hadn't had a pressing need to do so yet. But it will definitely be helpful for me to set that up at some point. Now seems like an appropriate time for me to learn how.

> Yes, you can use HTML tables if you like, it will work fine. You just can't mingle Markdown and HTML markup styles in the same block of text, really.

Great to know.

matildepark commented:
I sent an email about this in October:

1. Get the urbit.org repository cloned with zola and the submodule installed and ready to go. Genuinely, just do steps 1-4 from here: https://github.com/urbit/urbit.org
2. In the urbit.org repository, `cd content/docs`
3. `git checkout <your-working-branch>` (and maybe `git pull` to ensure it's the latest commit)
4. `cd ../..`
5. `zola serve`

Feel free to message if you have a hiccup; for a case like this, it would be best to have a live preview onhand.

syzygyzer commented:
Thanks, I definitely forgot about the email, but I started organizing incoming information much more pro-actively in recent months so hopefully these things will escape me less frequently now.

philipcmonk commented on reference/hoon-errors.md:
```diff
@@ -178,3 +187,388 @@ true:
+["||" %pass [%g %g] [[%deal [~zod ~zod] %dojo %poke] /use/hood/~zod/out/~zod/dojo/drum/phat/~zod/dojo] [i=/d t=~[//term/1]]]
+["|||" %give %g [%unto %fact] [i=/g/use/hood/~zod/out/~zod/dojo/drum/phat/~zod/dojo t=~[/d //term/1]]]
+["||||" %give %g [%unto %fact] [i=/d t=~[//term/1]]]
+["|||||" %give %d %blit [=//term/1 t=~]]
```
```suggestion
["|||||" %give %d %blit [i=//term/1 t=~]]
```

philipcmonk commented on reference/hoon-errors.md:
```diff
@@ -178,3 +187,388 @@ true:
+["||" %give %g [%unto %fact] [i=/g/use/dojo/~zod/out/~zod/spider/drum/wool t=~[/d //term/1]]]
+["|||" %give %g [%unto %fact] [i=/g/use/hood/~zod/out/~zod/dojo/drum/phat/~zod/dojo t=~[/d //term/1]]]
+["||||" %give %g [%unto %fact] [i=/d t=~[//term/1]]]
+["|||||" %give %d %blit [=//term/1 t=~]]
```
```suggestion
["|||||" %give %d %blit [i=//term/1 t=~]]
```

philipcmonk commented on reference/hoon-errors.md:
````diff
@@ -178,3 +187,388 @@ true:
+above:
+
+```
+["|||||" %give %d %blit [=//term/1 t=~]]
````
```suggestion
["|||||" %give %d %blit [i=//term/1 t=~]]
```

philipcmonk commented on reference/hoon-errors.md:
```diff
@@ -178,3 +187,388 @@ true:
+
+Let's put the first part of the stack trace into a table to make reading a little easier.
+
+| Length | move    | vane(s)   |                                                                                                     action | duct                                                                            |
```
I would associate the `%pass` wires with the either the `%pass` or the duct, probably the former.  Seems like a more natural place than with the action.  (Maybe this should be changed in the `|verb` output too, but out of scope for this)

philipcmonk commented on reference/hoon-errors.md:
```diff
@@ -178,3 +187,388 @@ true:
+This gives us a stack trace that is a list of `move`s and some
+associated metadata. Some of the `move`s are a bit of a distraction from what's
+going on overall such as acknowledgements that a `poke` was received
+(`%poke-ack`s), so we've omitted them for clarity. Furthermore, two `move`s (the
```
Good choice

syzygyzer commented:
We agreed where to put this material in the docs meeting today. I moved hoon-errors.md in its original form to the tutorial section at the top level, and the trace tutorial onto its own page in tutorials/arvo/move-trace.md, and made some minor revisions. Tomorrow I'll get urbit.org setup on my machine and fix the tables.

Putting hoon-errors.md at the top level means we now have 5 top level tutorials which is getting crowded, and I think we can still do a better job at organizing the tutorials, but that's for another day.

> A few small changes, and I would actually call it "event trace" or "kernel trace" or "|verb trace" or "move trace".

I changed "stack trace" to "move trace" in https://github.com/urbit/docs/pull/801/commits/647bcc00049ecac19cd8e891f2ec3138bc1c5c67 without really giving it much thought. I don't have convictions in any direction here other than agreeing that calling this a stack trace is overloading the term and so we should pick something else. I might have a stronger opinion tomorrow.

syzygyzer commented:
After much huffing and puffing, I put together a diagram using graphviz that visualizes the stack trace:

https://media.urbit.org/docs/arvo/move-trace.png

I'll work on a key tomorrow, but essentially, blue = %pass moves, which list the card, wire, and duct, and green = %give moves, which list the card and the duct.

I'd love some feedback on how to improve the diagram. I already have a couple of things in mind, like making boxes to separate the Unix stuff and the Arvo stuff. I squashed the Unix side of things to not include Vere, but it would be easy to add that in. Vere isn't mentioned in the tutorial right now, so I chose to gloss over that detail.

I put the pass/gives with Ford and Behn on the right hand side to emphasize the fact that its a sequence of pass/gives rather than a sequence of passes like most of the main column, but I could move that back in. It just makes the whole thing really tall and I think makes it a little more readable.

The information about depth (`||..|` at the beginning of each line of the trace) is implicit in the path length along the graph starting with the Arvo kernel, but another thought I had was to color code that with a gradient. Putting the bars in the label was a little ugly, but is another option.

Different arrow types denote whether something is a `note`, `task`, `gift`, or `sign`.

philipcmonk commented:
I don't see this on urbit.org, does this not get auto-deployed?

matildepark commented:
It does not, it's manual and I've been busier than usual. Let me push.

rmariani commented:
Thanks @matildepark!

matildepark commented:
I know it can be automated, and I've written a workflow proposal (last September) to create an automated, target-agnostic documentation deployment process — just hasn't been in the cards to enact that on my behalf. I haven't written webhook and bot stuff yet, heh.

philipcmonk commented:
Cool, no rush.  Didn't realize it's manual, that's fine.
